cloudsync/smartsync.py

```diff
@@ -0,0 +1,374 @@
+        self._root_paths: List[str] = list(root_paths) if root_paths else [None, None]
+        if not sleep:
+            # these are the event sleeps, but really we need more info than this
+            sleep = (self.providers[LOCAL].default_sleep, self.providers[REMOTE].default_sleep)
+
+        self.sleep = sleep
+
+        ####
+
+        max_sleep = max(sleep)                    # on sync fail, use the worst time for backoff
+        self.aging = max_sleep / 5                # how long before even trying to sync
+        self.min_backoff = max_sleep / 10.0       # event sleep of 15 seconds == 1.5 second backoff on failures
+        self.max_backoff = max_sleep * 10.0       # escalating up to a 3 minute wait time
+        self.mult_backoff = 2
+        assert len(self.providers) == 2
+    def pre_sync(self, sync: SyncEntry) -> bool:  # pylint: disable=too-many-branches
+        finished = super().pre_sync(sync)
+        local_file = sync[LOCAL].oid and self.providers[LOCAL].exists_oid(sync[LOCAL].oid)
+        if not finished:
+            finished = not (local_file or sync in self.state.requestset or sync[REMOTE].otype == DIRECTORY)
+        return finished
+
+    def get_parent_conflicts(self, sync: SyncEntry, changed) -> List[SyncEntry]:
+        """Returns list of parent conflicts."""
+        done = False
+        pcs = []
+        new_pc = sync
+        while not done:
+            pc = self._get_parent_conflict(new_pc, changed)
+            # if pc:  # TODO: test if this is necessary
+            #     pc.get_latest()
+            if pc and pc[REMOTE].path != new_pc[REMOTE].path:
+                pcs.append(pc)
+                new_pc = pc
+            else:
+                done = True
+        pcs.reverse()
+        return pcs
+
+
+class SmartSyncState(SyncState):
+    """Enhances the syncstate to support smart syncing."""
+    def __init__(self,
+                 providers: Tuple['Provider', 'Provider'],
+                 storage: Optional['Storage'] = None,
+                 tag: Optional[str] = None,
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset: Set[SyncEntry] = set()
+        self.excludeset: Set[SyncEntry] = set()
+        self._callbacks: List[Callable] = list()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def register_auto_sync_callback(self, callback):
+        self._callbacks.append(callback)
+
+    def _smart_sync_ent(self, ent):
+        if ent[LOCAL].path and not self.providers[LOCAL].exists_path(ent[LOCAL].path):
+            ent[LOCAL].clear()
+            ent[REMOTE].sync_path = None
+            ent[REMOTE].sync_hash = None
+            self.update_entry(ent, REMOTE, None, changed=True)
+        self.requestset.add(ent)
+        self.excludeset.discard(ent)
+
+    def smart_sync_path(self, remote_path) -> List[SyncEntry]:
+        # q: do we want to raise an exception if path isn't found?
```

Comment:
I think a `FileNotFoundError` makes sense

Reply:
done

Reply:
FYI, I changed all the `FileNotFoundError`'s to `CloudFileNotFoundError`'s everywhere in smartsync.py